sunscreen_runtime/src/debugger/server.rs

```diff
@@ -321,3 +322,42 @@ pub async fn get_stack_trace(
     }
     Ok(HttpResponse::NotFound().body(format!("Session {} not found", session)))
 }
+
+#[cfg(feature = "debugger")]
+#[get("sessions/{session}/groups/{groupid}")]
```

Comment:
I think this API should return something like

```rust
enum NodeData {
  Group(u64,) // GroupId
  FheOperation(FheOperation),
  ZkpOperation(ZkpOperation)
}

// Return this thing
StableGraph<NodeData, EdgeInfo>
```

This should be the graph to display for this group.

Reply:
You can construct this graph dynamically from the full graph by doing

```rust
let child_groups = id_data_lookup.children_for(groupid).iter().collect::<HashSet<u64>>();

let mut graph = match curr_session {
  Session::BfvSession(curr_session) => {
    curr_session.graph.filter_map(|_, n| {
      if n.group_id == groupid {
        Some(DisplayNodeInfo::FheOperation(n))
      }
    }, |_, e| Some(e))
  },
  Session::ZkpSession(curr_session) => {
    curr_session.graph.filter_map(|_, n| {
      if n.group_id == groupid {
        Some(DisplayNodeInfo::ZkpOperation(n))
      }
    }, |_, e| Some(e))
  }
};

// Map group ids to the node ids you add to the graph.
let child_groups = child_groups.iter().map(|g| (g, graph.add_node(DisplayNodeInfo::Group(g)))).collect::<HashMap<u64, NodeIndex>>();

// Iterate nodes we added and see if we need to hook them up to a group node
for n in graph.node_indices() {
  for e in graph.neighbors_directed(n, Direction::Outgoing) {
    if child_groups.contains_key(graph.node_weight(e.target()).unwrap().group_id) {
      graph.add_edge(n, e.target, DisplayEdgeInfo::GroupEdge);
    }
  }

  for e in graph.neighbors_directed(n, Direction::Incoming) {
    if child_groups.contains_key(graph.node_weight(e.source()).unwrap().group_id) {
      graph.add_edge(n, e.target, DisplayEdgeInfo::GroupEdge);
    }
  }
}
```

Reply:
That should replace all your complex data structures.